File: backend/src/Cupola.Core/Services/ObjectSaveResult.cs
using Cupola.Core.Models;

namespace Cupola.Core.Services;

/// <summary>
/// Per-object outcome vocabulary for <see cref="ObjectSaveResult"/> (B04).
/// Serialized as lower-case strings so the shared frontend type matches.
/// </summary>
public static class ObjectSaveOutcome
{
    public const string Created = "created";
    public const string Updated = "updated";
    public const string Conflict = "conflict";
}

/// <summary>
/// The outcome of persisting one domain object (OMCT-C04-L2-02.03,
/// OMCT-C04-L2-02.07). On success <see cref="Object"/> carries the saved
/// state; on conflict it carries the current stored state.
/// </summary>
public record ObjectSaveResult(string KeyString, string Outcome, DomainObject? Object);

File: frontend/projects/core/src/lib/models/object-save-result.ts
import { DomainObject } from './domain-object';

/**
 * B04 — per-object outcome of a persistence save (OMCT-C04-L2-02.03).
 * `conflict` marks an optimistic-concurrency rejection (OMCT-C04-L2-02.07).
 */
export type ObjectSaveOutcome = 'created' | 'updated' | 'conflict';

export interface ObjectSaveResult {
  keyString: string;
  outcome: ObjectSaveOutcome;
  /** The stored object: the saved state on success, the current server state on conflict. */
  object: DomainObject | null;
}
